Remove swarm inspect and use info instead

vdemeester opened the pull request:
Remove the `docker swarm inspect` command and use `docker info` instead to display swarm information if the current node is a manager 🐮.

This only change the cli side and the return of `/info` but not the `/swarm/inspect` endpoint. The _attribute_ used in the info endpoint is `Swarm.Cluster` and is populated only when the node is a manager.

```
$ docker info
# […]
Swarm: active
 NodeID: b1n3yqwdmvgpbhgmiv4xh48xy
 Is Manager: true
 ClusterID: d3aen0t7fkdw6kq1y4e3w4vv6
 Managers: 1
 Nodes: 1
 Name: default
 Orchestration:
  Task History Retention: 10
 Raft:
  Snapshot interval: 10000
  Heartbeat tick: 1
  Election tick: 3
 Dispatcher:
  Heartbeat period: 5 seconds
 CA configuration:
  Expiry duration: 3 months
# […]
```

I did not add all the thing displayed by `swarm inspect` but they are available through the API.
- [x] Complete attribute displayed ?
- [x] <del>Remove `/swarm/inspect` endpoint ?</del>

I'll create a `engine-api` if this pass design-review :wink: 

Closes #24148

/cc @aluzzardi @tiborvass @stevvooe @tonistiigi 

Signed-off-by: Vincent Demeester vincent@sbr.pm

AkihiroSuda commented on api/client/system/info.go:
```diff
@@ -87,6 +87,26 @@ func runInfo(dockerCli *client.DockerCli) error {
+			fmt.Fprintf(dockerCli.Out(), " Acceptance Policies:\n")
+			for _, policy := range info.Swarm.Cluster.Spec.AcceptancePolicy.Policies {
+				fmt.Fprintf(dockerCli.Out(), "  %s:\n", policy.Role)
+				fmt.Fprintf(dockerCli.Out(), "   auto-accept: %t\n", policy.Autoaccept)
```
I feel `Auto-acceptance` and `Secret` are better for consistency

vdemeester replied:
@AkihiroSuda you're right :+1:

icecrime commented:
A few remarks about acceptance policies:
- I don't think we should display the secret: people copy/paste this in GitHub issues :innocent: 
- We can probably condense the output (e.g., `Auto-accept: (manager|worker|none)`).

vdemeester commented:
Rebased and hopefully, tests fixed 👼

@icecrime 

> I don't think we should display the secret: people copy/paste this in GitHub issues :innocent: 

`docker swarm inspect` did display it but I'm ok with not displaying in `docker info` (it will still be present in the json)

> We can probably condense the output (e.g., `Auto-accept: (manager|worker|none)`).

That's an idea, I'll look into it.

vdemeester commented:
Hum, removing `/swarm/inspect`means removing `client.SwarmInspect` (from `engine-api`) and means making `swarm update` call the `/info` endpoint instead. 

I'm not sure we should remove it, just adding *Cluster* information in `/info`.

@thaJeztah @icecrime wdyt ?

thaJeztah commented:
Perhaps we can keep the  `inspect` endpoint for now, if that simplifies things

tiborvass commented:
@vdemeester sorry but would you mind linking where `swarm update` calls `/swarm` (inspect) ?

vdemeester commented:
@tiborvass yes, here https://github.com/docker/docker/blob/master/api/client/swarm/update.go#L35 👼

icecrime commented:
I think it's ok for `swarm update` to call the `/info` endpoint as long as the `SwarmSpec` is a nested element of the result, and not something that I need to construct myself for the purpose of calling `/swarm/update`.

icecrime commented:
Also :wink: 

```
11:22:56 FAIL: docker_cli_swarm_test.go:15: DockerSwarmSuite.TestSwarmUpdate
11:22:56 
11:22:56 [d66998809] waiting for daemon to start
11:22:56 [d66998809] daemon started
11:22:56 docker_cli_swarm_test.go:24:
11:22:56     c.Assert(err, checker.IsNil, check.Commentf("out: %v", out))
11:22:56 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc822f77080), Stderr:[]uint8(nil)} ("exit status 1")
11:22:56 ... out: Error response from daemon: page not found
```

vdemeester commented:
@icecrime ok, I'll update and prepare a PR for engine-api. The test failure is related to that :wink:

vdemeester commented:
Updated, fixed test failures and created a PR in `engine-api` (https://github.com/docker/engine-api/issues/318) — vendor is expected to fail 👼.

tiborvass commented:
@vdemeester let's keep the API unchanged and have `inspect` make multiple calls for now. Not sure if you would still need #318 then.

vdemeester commented:
@tiborvass you mean `info` 👼

vdemeester commented:
Updated 👼

vieux commented:
@vdemeester `docker_cli_swarm_test.go` uses `docker swarm inspect`

vdemeester commented:
@vieux ah damn.. revert the wrong file… I'll fix 👼

tiborvass commented:
@vdemeester some updates we need on this one:

- Rebase :)
- Update description to show the new information, since the result from the swarm API has changed (no more acceptance policy, CAhash, etc.)
- Make sure we DO NOT display join tokens (because they are returned by the API)
- Make only 1 API call to /info, and gather all the swarm information server side.
- Having `Cluster:` and `Swarm:` makes no sense, they are synonyms, they should be merged, and `ID` be renamed to `ClusterID`.
- Remove `Version` from `Cluster:`
- `Heartbeat period` should use human formatting for duration

kencochrane commented:
Can we add an `IsLeader` field to show if this node is the lead manager. I'm currently using swarm inspect for this, but if swarm inspect is going away, then I need another way to get that value.

Another question. Will `Cluster:Id` be available on the worker nodes as well as the manager nodes? It isn't today and that would be very helpful. It is an easy way to see what swarm a node is in for logging and analytics/metrics gathering for nodes by cluster.

vdemeester commented:
> Make only 1 API call to /info, and gather all the swarm information server side.

Ok this mean I need to make a PR in engine-api 👼

aluzzardi commented:
Hey Ken, regarding IsLeader, I think you can grab that from docker node
inspect self -f "{{ something something }}"

I do agree it'd be useful in docker info regardless so +1 for that request

On Saturday, July 23, 2016, Vincent Demeester <notifications@github.com>
wrote:

> Make only 1 API call to /info, and gather all the swarm information server
> side.
>
> Ok this mean I need to make a PR in engine-api 👼
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/24492#issuecomment-234730017>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAK9Lw_GdCh3gr-72PmrXCud3YXN8VH7ks5qYk-CgaJpZM4JI3dy>
> .
>

kencochrane commented:
@aluzzardi thanks for the work around. The one issue with `docker node` is that it only works for manager nodes, where as docker info works for all mode types.

aluzzardi commented:
Indeed - but for your use case, you can assume that if it's not a manager
node then it's probably not the leader :)

However I do still agree with you that this information should be available
on docker info - just pointing out a workaround

On Saturday, July 23, 2016, Ken Cochrane <notifications@github.com> wrote:

> @aluzzardi <https://github.com/aluzzardi> thanks for the work around. The
> one issue with docker node is that it only works for manager nodes, where
> as docker info works for all mode types.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/24492#issuecomment-234741369>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAK9L3EO3jswn-gSR1qJqmr0kjvx2HJpks5qYodVgaJpZM4JI3dy>
> .
>

vdemeester committed:
Remove swarm inspect and use info instead

Remove the swarm inspect command and use docker info instead to display
swarm information if the current node is a manager.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

tiborvass commented:
@vdemeester I was thinking about not returning join tokens via the /info API at all not just CLI

tiborvass commented:
Carried in #25042
